### PR TITLE
8301219: JavaFX crash when closing with the escape key

### PR DESCRIPTION
**Pull request by beldenfox**
While processing a key down event the Glass GTK code sends out PRESSED and TYPED KeyEvents back to back. If the stage is closed during the PRESSED event the code will end up referencing freed memory while sending out the TYPED event. This can lead to intermittent crashes.

In GlassApplication.cpp the EventCounterHelper object ensures the WindowContext isn't deleted while processing an event. Currently the helper object is being created *after* IME handling instead of before. If the IME is enabled it's possible for the WindowContext to be deleted in the middle of executing a number of keyboard-related events.

The fix is simple; instantiate the EventCounterHelper object earlier. There isn't always a WindowContext so I tweaked the EventCounterHelper to do nothing if the context is null.

To make the crash more reproducible I altered the WindowContext such that when it's deleted the freed memory is filled with 0xCC. This made the crash more reproducible and allowed me to test the fix. I did the same with GlassView since that's the only other Glass GTK class that's instantiated with `new` and discarded with `delete`.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issues
 * [JDK-8301219](https://bugs.openjdk.org/browse/JDK-8301219): JavaFX crash when closing with the escape key (**Bug** - P4)
 * [JDK-8087368](https://bugs.openjdk.org/browse/JDK-8087368): java runtime environment error when trying to execute showAndWait() function (**Bug** - P3)

### Reviewers
 * [Kevin Rushforth](https://openjdk.org/census#kcr) (@kevinrushforth - **Reviewer**)
 * [Thiago Milczarek Sayao](https://openjdk.org/census#tsayao) (@tsayao - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/1307/head:pull/1307` \
`$ git checkout pull/1307`

Update a local copy of the PR: \
`$ git checkout pull/1307` \
`$ git pull https://git.openjdk.org/jfx.git pull/1307/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1307`

View PR using the GUI difftool: \
`$ git pr show -t 1307`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/1307.diff">https://git.openjdk.org/jfx/pull/1307.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jfx/pull/1307#issuecomment-1854937097)

### CODE REVIEW

**Review by kevinrushforth — COMMENTED** [review]
The fix looks good. I tested it by reverting your fix and leaving in the debug memory fill on delete. The system test added by this PR crashes without the fix and passes with the fix. So do the tests associated with the fixed bugs.

I also ran a headful test run on our lab systems. All good.

As for the debug fill-on-delete code, it seems useful to have it available, but not enabled in production. I recommend to ifdef it out on a flag, either DEBUG or some unique flag defined in `DeletedMemDebug.h`, but off by default.

**Review by kevinrushforth — APPROVED** [review]
LGTM now. I did ask one question and will reapprove if you make the change.

File: modules/javafx.graphics/src/main/native-glass/gtk/DeletedMemDebug.h [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+public:
+    static void operator delete(void* ptr, std::size_t sz)
+    {
+        ::memset(ptr, FILL, sz);
+        ::operator delete(ptr);
+    }
+
+    static void operator delete[](void* ptr, std::size_t sz)
+    {
+        ::memset(ptr, 0xcc, sz);
```
[the remark this answers is not in the document]
Comment (beldenfox) [reply modules/javafx.graphics/src/main/native-glass/gtk/DeletedMemDebug.h]: Yes, it should use FILL. Will fix.

**Review by kevinrushforth — APPROVED** [review]

**Comment by kevinrushforth** [comment]
@tsayao Would you like to review this? If not, can someone else be the second reviewer?

**Comment by tsayao** [comment]
It's always failing on my machine - I'll check why.

**Comment by tsayao** [comment]
The change looks correct to me - the change makes sense, but I can't reproduce the crash on my systems.
I can confirm that it's reaching the  `delete ctx` correctly.

I didn't know about the `0xCC` debug trick.

**Review by tsayao — APPROVED** [review]
Looks good.

**Pull request closed without merging**